worker/undertaker/undertaker.go

```diff
@@ -102,27 +110,99 @@ func (u *Undertaker) Wait() error {
 }
 
 func (u *Undertaker) run() error {
-	result, err := u.config.Facade.ModelInfo()
-	// If model already gone, exit early.
-	if errors.IsNotFound(err) {
+	modelWatcher, err := u.config.Facade.WatchModel()
+	if errors.Is(err, errors.NotFound) {
+		// If model already gone, exit early.
 		return nil
+	} else if err != nil {
+		return errors.Trace(err)
 	}
+	err = u.catacomb.Add(modelWatcher)
 	if err != nil {
-		return errors.Trace(err)
+		return err
+	}
+	defer modelWatcher.Kill()
+
+	select {
+	case <-modelWatcher.Changes():
+	case <-u.catacomb.Dying():
+		return u.catacomb.ErrDying()
 	}
-	if result.Error != nil {
+
+	result, err := u.config.Facade.ModelInfo()
+	if errors.Is(err, errors.NotFound) {
+		// If model already gone, exit early.
+		return nil
+	} else if err != nil {
+		return errors.Trace(err)
+	} else if result.Error != nil {
 		return errors.Trace(result.Error)
 	}
-	modelInfo := result.Result
+	info := result.Result
 
-	u.config.Logger.Debugf(
-		"destroying model %q with timeout %v, force=%v",
-		modelInfo.Name, modelInfo.DestroyTimeout, modelInfo.ForceDestroyed)
+	ctx, cancel := context.WithCancelCause(context.Background())
+	defer cancel(context.Canceled)
 
-	if modelInfo.Life == life.Alive {
+	// Watch for changes to model destroy values, if so, cancel the context
+	// and restart the worker.
+	err = u.catacomb.Add(worker.NewSimpleWorker(func(stopCh <-chan struct{}) error {
+		for {
+			select {
+			case <-stopCh:
+				cancel(u.catacomb.ErrDying())
+				return nil
+			case <-modelWatcher.Changes():
+				result, err := u.config.Facade.ModelInfo()
+				if errors.Is(err, errors.NotFound) || err != nil || result.Error != nil {
+					continue
+				}
+				updated := result.Result
+				changed := false
+				switch {
+				case info.DestroyTimeout == nil && updated.DestroyTimeout != nil:
+					changed = true
+				case info.DestroyTimeout != nil && updated.DestroyTimeout == nil:
+					changed = true
+				case info.DestroyTimeout != nil && updated.DestroyTimeout != nil && *info.DestroyTimeout != *updated.DestroyTimeout:
+					changed = true
+				case info.ForceDestroyed != updated.ForceDestroyed:
+					changed = true
+				}
+				if changed {
+					u.config.Logger.Infof("model destroy parameters changed: restarting undertaker worker")
+					return errors.Errorf("model destroy parameters changed")
+				}
+			}
+		}
+	}))
+	if err != nil {
+		return err
+	}
+
+	if info.Life == life.Alive {
 		return errors.Errorf("model still alive")
 	}
-	if modelInfo.Life == life.Dying {
+
+	if info.ForceDestroyed && info.DestroyTimeout != nil {
+		u.config.Logger.Infof("force destroying model %q with timeout %v", info.Name, info.DestroyTimeout)
+		return u.forceDestroy(ctx, info)
+	} else if info.DestroyTimeout != nil {
+		u.config.Logger.Warningf("timeout ignored for graceful model destroy")
+	}
+	// Even if ForceDestroyed is true, if we don't have a timeout, we treat them the same
+	// as a non-force destroyed model.
```

Comment:
Shouldn't we catch this and raise an error. Allowing client code / CLI to submit what they believe is --force but then does a clean destroy is rather disingenuous. We should be very strict about ensuring only valid param combinations are passed given the number of permutations and real potential for confusion.

Reply:
--force without --timeout is a perfectly valid default, it does still force, but it waits for cleanup ops/tear down dance to finish first.

The undertaker is just there to ensure the cloud resources are indeed destroyed + the model is removed from state.